// my-util/models.js
/*
 * various mongoose models used in connection with the db
 * */
const mongoose = require("mongoose");

/* petbook.Users model */
const user_schema = mongoose.Schema({
    firstName: String,
    lastName: String,
    email: String,
    password: String,
    type: String,
    gender: String,
    birthday: String,
    friends: [String],
    avatar: String
});
const user_model = mongoose.model("user", user_schema, "Users");

/* petbook.Photos model */
const photo_schema = mongoose.Schema({
    path: String,
    owner: String
});
const photo_model = mongoose.model("photo", photo_schema, "Photos");

/*
 * petbook.Posts model
 *
 * for now also store avatar id in here (this can be obtained from uid)
 * since it's not that bad to duplicate ids anyway
 * this makes it easier to show user avatar near each post (saves one
 * db query)
 * */
const post_schema = mongoose.Schema({
    uid: String,
    text: String,
    timestamp: String,
    avatar: String,
    comments: [String]
});
/* TODO first arg? */
const post_model = mongoose.model("post", post_schema, "Posts");

/* exports */
module.exports.user_model = user_model;
module.exports.photo_model = photo_model;
module.exports.post_model = post_model;
