// my-util/photos.js
/*
 * module for interacting with photos collection
 */

const EventEmitter = require("events");

const default_avatar_id = "59fcd27c192bce16b2465bef";
const default_avatar_path = "photos/default_avatar.png";

const connect = require("./connect.js");
const models = require("./models.js");

function get_photo_url(photo_id, ee, event) {
    /*
     * returns photo path/url for photo id
     * */
    const photo_emitter = new EventEmitter();
    photo_emitter.on("connection", function(ret) {
	if (ret["success"] === false) {
	    ee.emit(event, ret);
	    return;
	}
	let filter = {_id: photo_id};
	models.photo_model.find(filter, function(err, result){
	    if (err){
		let ret = {
		    success: false,
		    messages: ["Error retrieving photos: " + err.message]
		};
		ee.emit(event, ret);
		return;
	    }
	    if (result.length < 1){
		let ret = {
		    success: false,
		    messages: ["Could not find photo id"]
		};
		ee.emit(event, ret);
		return;
	    }
	    let ret = {
		success: true,
		messages: ["success"],
		photo_url: result[0]["path"]
	    };
	    ee.emit(event, ret);
	});

    });
    /* wait for connection if not opened */
    if (!connect.connection_active())
	connect.connect_to_db(photo_emitter, "connection");
    else
	photo_emitter.emit("connection", {success: true});
}

/* exports */
module.exports.default_avatar_id = default_avatar_id;
module.exports.default_avatar_path = default_avatar_path;
module.exports.get_photo_url = get_photo_url;
