// my-util/users.js
/*
 * module for interacting with users collection
 */

const EventEmitter = require("events");

const connect = require("./connect.js");
const models = require("./models.js");

function get_users(ee, event) {
    /*
     * function that will retrive all info for all users
     * */

    /* first, register callbacks */
    const users_emitter = new EventEmitter();

    /* after connecting to db */
    users_emitter.on("connection", function(ret){
	if (ret["success"] === false) {
	    ee.emit(event, ret);
	    return;
	}
	/* TODO do we need empty filter here? */
	models.user_model.find({}, function(err, result){
	    if (err){
		let ret = {
		    success: false,
		    messages: ["Error retrieving users: " + err.message],
		    users: []
		};
		ee.emit(event, ret);
		return;
	    }
	    /*
	     * I guess there will always be users
	     * since you need to be logged in to use this
	     * */
	    let users = [];
	    result.forEach(function(elem){
		users.push({
		    id: elem._id,
		    firstName: elem.firstName,
		    lastName: elem.lastName,
		    email: elem.email,
		    password: elem.password,
		    type: elem.type,
		    gender: elem.gender,
		    birthday: elem.birthday,
		    friends: elem.friends,
		    avatar: elem.avatar
		});
	    });
	    let ret = {
		success: true,
		messages: ["success"],
		users: users
	    };
	    ee.emit(event, ret);
	});
    });

    /* connect to db */
    connect.connect_to_db(users_emitter, "connection");
}

function update_avatar(data, ee, event) {
    /*
     * function that will update avatar for user
     *
     * data: {uid, photo_id}
     * */
    const user_emitter = new EventEmitter();

    /* when connection is done */
    user_emitter.on("connection", function(ret){
	if (ret["success"] === false) {
	    ee.emit(event, ret);
	    return;
	}
	let filter = {_id: data.uid};
	let updated_values = {
	    avatar: data.photo_id
	};
	models.user_model.update(filter, updated_values, function(err, result){
	    if (err){
		let ret = {
		    success: false,
		    messages: ["Error updating avatar: " + err.message]
		};
		ee.emit(event, ret);
		return;
	    }
	    let ret = {
		success: true,
		messages: ["success"]
	    };
	    ee.emit(event, ret);
	});
    });
    /* connect to db */
    connect.connect_to_db(user_emitter, "connection");
}

function get_user_info(uid, ee, event) {
    /*
     * function that will retrive all info for user afer id
     *
     * on failure all fields/props will be undefined
     * */

    /* first, register callbacks */
    const user_emitter = new EventEmitter();

    /* after connecting to db */
    user_emitter.on("connection", function(ret){
	if (ret["success"] === false) {
	    ee.emit(event, ret);
	    return;
	}
	let filter = {_id: uid};
	models.user_model.find(filter, function(err, result){
	    if (err){
		let ret = {
		    success: false,
		    messages: ["Error retrieving user: " + err.message]
		};
		ee.emit(event, ret);
		return;
	    }
	    if (result.length < 1) {
		let ret = {
		    success: false,
		    messages: ["User not found"]
		};
		ee.emit(event, ret);
		return;
	    }
	    /*
	     * there should be some privacy concerns here
	     * for now, return everything, but whoever calls this
	     * should show/hide info based on privacy
	     * */
	    let ret = {
		success: true,
		messages: ["success"],
		uid: result[0]["_id"],
		firstName: result[0]["firstName"],
		lastName: result[0]["lastName"],
		email: result[0]["email"],
		password: result[0]["password"],
		type: result[0]["type"],
		gender: result[0]["gender"],
		birthday: result[0]["birthday"],
		friends: result[0]["friends"],
		avatar: result[0]["avatar"]
	    };
	    ee.emit(event, ret);
	});
    });

    /* connect to db */
    connect.connect_to_db(user_emitter, "connection");
}

/* exports */
module.exports.get_user_info = get_user_info;
module.exports.update_avatar = update_avatar;
module.exports.get_users = get_users;
